### DISCUSSION TITLE
Queries regarding Google's experience of TNA and other orchestration/management platforms for their Content Data Networks

### DISCUSSION THREAD

**Discussion by varada-samsung** [nephio-project/nephio#442] [Development]
1) Are there any provisions for finding the health of the system in the form of Alerts ( for alarms, events ), KPIs of a particular workload ( beyond the generic CPU/Memory of workload ), etc.
 2) If the answer to 1) is Yes, what are the components involved? 
3) Any performance baseline for the components identified in 2) like X alerts per second processed etc.? 
4) Are there any "Operators" built and reconciled successfully more than what is offered in basic K8 operators like pod status?  For example, a workload processing rate is X transactions per second and if it is below that, spawn a new workload (or) configure the existing workload to increase the transactions rate
 5) Does 4) involve any AI/ML component?
 6) In the "Overview of Nephio" - Figure 3: Reference Implementation and Google Open Source - https://wiki.nephio.org/display/HOME/Overview+of+Nephio, there was a mention of Metric Service to be taken from an existing open source.  What are those components?  Are they the same or modified verison of K8 metric service? 
8. How are workloads running in GCP configured?
9. How generic an INTENT is and few examples of a practical INTENT?
10. Upto what level an INTENT can go deep? Upto, workload level? Application level?

**Comment by johnbelamaric** [comment]
@tliron @s3wong @kaushikgoa-google perhaps you can provide some insight?

**Comment by tliron** [comment]
> 1. Are there any provisions for finding the health of the system in the form of Alerts ( for alarms, events ), KPIs of a particular workload ( beyond the generic CPU/Memory of workload ), etc.

This topic is still in its infancy.

1) There is a new taskforce on observability for SIG#2. At this point they are still collating requirements and evaluating technologies.
2) Google seed code for Nephio includes a status aggregation framework. Integrating it into a Nephio release is still ongoing.

>     3. Any performance baseline for the components identified in 2) like X alerts per second processed etc.?

Not at present. I have proposed a scalability taskforce that would look at this as well as other bandwidth and storage scalability topics.

>     4. Are there any "Operators" built and reconciled successfully more than what is offered in basic K8 operators like pod status?  For example, a workload processing rate is X transactions per second and if it is below that, spawn a new workload (or) configure the existing workload to increase the transactions rate

This topic is a high priority for me personally. In Nephio R1 and R2 we have not shown this ability quite yet, as our scaling/healing scenarios are handled in the management cluster based on user/external triggers. The best place to develop these capabilities is likely in the ongoing work on a Service Assurance use case, in the SIG#1 workgroup. Achieving this could involve proper observability (see above) as well as integration with a policy framework, such as Open Policy Agent (OPA). However, my personal opinion is that we can move forward with a PoC even if we do not have clear technology choices yet. I will propose such a PoC for Nephio R3.

>     5. Does 4) involve any AI/ML component?

Not at present, though there is interest in this and potential for contribution from Google and other parties.

>     6. In the "Overview of Nephio" - Figure 3: Reference Implementation and Google Open Source - https://wiki.nephio.org/display/HOME/Overview+of+Nephio, there was a mention of Metric Service to be taken from an existing open source.  What are those components?  Are they the same or modified verison of K8 metric service?

As stated above, this is the purview of the observability taskforce. The general intent in Nephio is to not reinvent wheels but to rely on "bread and butter" technologies from CNCF, LFN, and the greater Kubernetes ecosystem.

> >     8. How are workloads running in GCP configured?

"Configuration" is front and center for Nephio. Specifically let's divide it to what I call "pre-stand-up" and "post-stand-up" configuration. So far Nephio has been focused on the former, in the [Configuration as Data (CaD)](https://cloud.google.com/blog/products/containers-kubernetes/understanding-configuration-as-data-in-kubernetes) paradigm. More specifically, network function packaging provides a unified frontend for both network function *and* infrastructure requirements. This allows for controllers to configure the infrastructure before and during (Day 2) the workload operation.

>      9. How generic an INTENT is and few examples of a practical INTENT?

As stated, this has been the core of Nephio development thus far. The best way to appreciate the scope is by going through our example use cases.

>      10. Upto what level an INTENT can go deep? Upto, workload level? Application level?

Yes, yes, and yes. :) As stated, the focus is on the special challenge that network functions are strongly vertically integrated between software and hardware, with the cloud infrastructure (Kubernetes, CNI, Linux OS and drivers, GPU/DPU, ToR switches, etc.) in between.

I would interpret "application level" here as what I called "post-stand-up configuration", e.g. NETCONF and gNMI interactions with the network function *after* it has been deployed. We have not addressed this yet, but there is a lot of interest in this area and we plan some integrated solutions in the future.
